runtime/buffer.h

```diff
@@ -0,0 +1,179 @@
+//===-- runtime/buffer.h ----------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// External file buffering
+
+#ifndef FORTRAN_RUNTIME_BUFFER_H_
+#define FORTRAN_RUNTIME_BUFFER_H_
+
+#include "io-error.h"
+#include "memory.h"
+#include <algorithm>
+#include <cinttypes>
+#include <cstring>
+
+namespace Fortran::runtime::io {
+
+void LeftShiftBufferCircularly(char *, std::size_t bytes, std::size_t shift);
+
+// Maintains a view of a contiguous region of a file in a memory buffer.
+// The valid data in the buffer may be circular, but any active frame
+// will be contiguous.  The requirement is to preserve read data that
+// may be reused by means of Tn/TLn edit descriptors without needing to
+// position the file (which may not always be possible, e.g. a socket),
+// and as a general objective to reduce system call counts.
+//
+// This CRTP mixin maintains a circular buffer of valid data that has been
+// read from or is yet to be written to a file.  Within that interval of
+// valid data, it supports a current frame that is guaranteed to be
+// contiguous.  The API comprises routines that establish a given interval
+// of the file as the current frame for reading or writing,
+template<typename STORE> class FileFrame {
+public:
+  using FileOffset = std::uint64_t;
+
+  ~FileFrame() { FreeMemoryAndNullify(buffer_); }
```

Comment:
When I build using GNU 8.3.0, I get an error message for this line:
../runtime/buffer.h:40:39: error: cannot bind non-const lvalue reference of type â€˜void*&â€™ to an rvalue of type 'void*'
It builds OK for me with Clang.

Reply:
Fixed.